feat: [M3-7129] Update PR template

abailly-akamai opened the pull request:
## Description 📝
This PR updates our template when opening a PR. I welcome all suggestions to improve it as this is something we use every day and all want to be in (relative) agreement about.

It does streamline the previous iteration by focusing on shipping quality rather than the kind of commit we expect, which is now an obsolete concept with changesets.

## Preview 📷
| Before  | After   |
| ------- | ------- |
| ![Screenshot 2023-10-02 at 15 21 49](https://github.com/linode/manager/assets/130582365/54a46357-cca9-4b6e-856d-f83008643174) | ![Screenshot 2023-10-06 at 11 08 28](https://github.com/linode/manager/assets/130582365/3bd1d69d-76f9-4c2d-9544-f4095b04c803) |

## How to test 🧪
1. If wanting to look at markup rendition, either look at an online converter such as http://demo.showdownjs.com/ or pull the code locally and use the MD preview from VScode. None of those solutions will exactly mirror the final result but will help get a sense of the final result.

abailly-akamai commented on docs/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -1,45 +1,47 @@
-## Description 📝
-**Brief description explaining the purpose of the changes**
```
still relevant and moved over

abailly-akamai commented on docs/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -1,45 +1,47 @@
-## Major Changes 🔄
-**List highlighting major changes**
-- Change #1
-- Change #2
```
relevant (now as part of the description)

mjac0bs replied:
I did like this as its own section (and just delete it in my own PRs if it's not relevant), but would be okay with deleting it as long as we make the below change to Description. ⬇️

abailly-akamai replied:
I think it is an ambiguous and i'd like to reword it then. What makes a change major? and should i not list minor changes? The nomenclature also seems to refer to semver (major/minor/patch) and could be misinterpreted that way (that was my experience)

mjac0bs replied:
All fair points.

> The nomenclature also seems to refer to semver (major/minor/patch) and could be misinterpreted that way (that was my experience)

We could replace "major" with "significant" or "important" to address this issue, but that doesn't really address your other concerns. 

> What makes a change major? and should i not list minor changes? 

I'm not sure I have a great answer and am curious what other team members' thoughts are on our template, but when there is good acceptance criteria for a ticket, I happily copy it over and include those as my "major changes", plus anything else I subjectively think is important (e.g. if I found and fixed a minor bug). I often find it easier to read a PR's changes in list form, so I'll include minor things rather than write a wall of text as a description.

coliu-akamai replied:
I like the Description + List of changes!

hmm in terms of things I normally put in the major changes section, I tended to include:
- acceptance criteria (definitely did this more for the MUI migrations though)
- large (which I guess is also subjective 😅) changes like new components, new files, new functions
- bug fixes if that ended up being a part of the PR but wasn't in the original ticket description

abailly-akamai commented on docs/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -1,45 +1,47 @@
 
 | Before  | After   |
 | ------- | ------- |
-| Content | Content |
```
Moved over as well

abailly-akamai commented on docs/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -1,45 +1,47 @@
-1. **How to setup test environment?**
-2. **How to reproduce the issue (if applicable)?**
-3. **How to verify changes?**
-4. **How to run Unit or E2E tests?**
```
now broken down between prerequisites and actual steps

tyler-akamai replied:
From how I read it, I think there might be a section missing:
- Prerequisites - How to setup test environment
- Reproduction steps - How to reproduce the issue (if applicable)
- Verification steps (missing) - How to verify changes

tyler-akamai replied:
and I think the unit tests and E2E tests can just be included in the Verification steps. 

Thats my opinion, other people might have different ones.

abailly-akamai commented on docs/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -1,45 +1,47 @@
-- `test`: New tests or changes to existing tests. Does not change the production code.
-- `change`: Only use for something that doesn’t fit in any other category.
-
-**Example:** `feat: [M3-1234] - Allow user to view their login history`
```
We do not need to format our commit this was anymore since switching over to changesets.

mjac0bs replied:
While no longer necessary for the changelog, I do really like that our PR titles are consistent -- it makes the list of them easy to skim and get a sense of what it contains just from looking at the title. I think we should keep this list as a reminder for those of us on the team and especially as a helpful guide for contributors outside our team.

abailly-akamai replied:
That's fine, but can we streamline it?  it's very long and no one reads it (hopefully our one contributor a year does 😆).

at the very least it should match out PR types.

abailly-akamai replied:
this was re-added but streamlined to mirror our PR types

carrillo-erik replied:
I thought the commit types helped with informing the type of work the PR covered. This change also eliminates some of the other commit types. For example, this PR should be `docs: [M3-7129] - Update PR Template` under the old guidelines.

mjac0bs commented on docs/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -1,45 +1,47 @@
 
-## Preview 📷
+## Description
+Highlight the Pull Requests context and intentions and provide a list of major changes.
```
```suggestion
Highlight the Pull Request's context and intentions.
Provide a list of major changes.
```
I'm hoping that by separating the two requests into two sentences, it will remind us to do both these things.

abailly-akamai replied:
same comment on the "Major" changes

mjac0bs commented on docs/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -1,45 +1,47 @@
+- ...
+- ...
+
+### Post deployment tasks
```
I'm not sure I understand what this means.

abailly-akamai replied:
All of our PRs end up in development and can sit in there for x days before ending up in production. This could be a note to the reviewer that something else needs to happen (like turning on a flag) once in production.

We can def lose this, you're right it's not very clear

carrillo-erik commented:
I think the PR template could also mention running the `changeset` script and including it as part of the PR.

carrillo-erik commented on docs/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -1,45 +1,64 @@
+## As an Author I have considered
+*Check all that apply*<br>
+<br>
+[ ] 👀 Doing a self review<br>
```
I liked having the emoji on the section headers, I found it easier to identify each section. I'm fine with these emojis, but maybe we could also keep them in the section headers?

abailly-akamai replied:
added back - also the PR template does already mention adding a changeset

carrillo-erik commented:
@abailly-akamai  Is there a separate ticket to update the `CONTRIBUTING.md` file? We have a whole section in there named `Submitting a pull request`. I think it would benefit to sync the two docs once this one is finalized.

dwiley-akamai commented on docs/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -1,45 +1,64 @@
 **Include a screenshot or screen recording of the change**
 
-> **Note**: Use `<video src="" />` tag when including recordings in table
+:bulb: Use `<video src="" />` tag when including recordings in table
```
Doesn't look like `:bulb:` gets turned into an emoji 🤔

dwiley-akamai commented on docs/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -1,45 +1,64 @@
+[ ] Bug Fix<br>
+[ ] Change<br>
+[ ] Tech Story / Refactor<br>
+[ ] Upcoming Feature
```
Do we need this section if the PR title is formatted properly?

abailly-akamai replied:
Fair enough, since I added back the formatting rule i removed this

abailly-akamai commented on docs/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -1,45 +1,64 @@
-- `change`: Only use for something that doesn’t fit in any other category.
-
-**Example:** `feat: [M3-1234] - Allow user to view their login history`
+- `upcoming`: A new feature that is in progress, not visible to users yet, and usually behind a feature flag.
```
I still think those should correspond to our changeset types

abailly-akamai replied:
chore can be replaced by a more descriptive type available

abailly-akamai commented:
@linode/frontend Addressed feedback and looking to get this merged.

The people have spoken. In the end it's very few changes
- adding the laundry list section
- narrowed our commit types to match our changeset types (both in the template and the contributing document)
- Few copy changes to clarify things or remove ambiguity 

Can't please everyone please give it another look and we can also iterate after using it for a bit. thx!

jdamore-linode reviewed (approved):
Thanks @abailly-akamai!

I'm a fan of these changes, but also happy to keep iterating if we discover there's more room for improvement 👍

jcallahan-akamai reviewed (approved):
Looks good, left one suggestion.

bnussman-akamai commented on docs/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -1,45 +1,68 @@
+- ...
+- ...
+
+## As an Author I have considered 🤔
+*Check all that apply*<br>
+<br>
+[ ] 👀 Doing a self review<br>
+[ ] ❔ Our [contribution guidelines](https://github.com/linode/manager/blob/develop/docs/CONTRIBUTING.md)<br>
+[ ] 🤏 Splitting feature into small PRs<br>
+[ ] ➕ Adding a changeset<br>
+[ ] 🧪 Providing/Improving test coverage<br>
+[ ] 🚩 Using a feature flag to protect the release<br>
+[ ] 👣 Providing comprehensive reproduction steps<br>
+[ ] 📑 Providing or updating our documentation<br>
+[ ] 🕛 Scheduling a pair reviewing session<br>
+[ ] 📱 Providing mobile support<br>
+[ ] ♿  Providing accessibility support
```
Just a general question: Is it intended that we keep this visible in the PR description when we publish it or is it just a checklist for our own sake that is meant to be omitted after we verify all of the items

abailly-akamai replied:
I feel like it could be helpful to the reviewer as well. For instance, you may check that you added a changeset but you had actually forgotten to do it so the reviewer can catch that

abailly-akamai commented on docs/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -1,45 +1,68 @@
+[ ] 🕛 Scheduling a pair reviewing session<br>
+[ ] 📱 Providing mobile support<br>
+[ ] ♿  Providing accessibility support
+
```
sure thing, added!

coliu-akamai reviewed (approved):
thank you Alban! 🎉
